assets/dns/daemonset.yaml

```diff
@@ -89,18 +89,23 @@ spec:
           while true; do
             declare -A svc_ips
             for svc in "${services[@]}"; do
-              # Fetch service IP from cluster dns if present
-              ips=($(dig -t A @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"))
-
-              if [[ "$?" -eq 0 && "${#ips[@]}" -ne 0 ]]; then
-                svc_ips["${svc}"]="${ips[@]}"
-              else
-                # Try IPv6
-                ips=($(dig -t AAAA @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"))
+              # Fetch service IP from cluster dns if present. We make several tries
+              # to do it: IPv4, IPv6, IPv4 over TCP and IPv6 over TCP. The two last ones
+              # are for deployments with Kuryr on older OpenStack (OSP13) - those do not
+              # support UDP loadbalancers and require reaching DNS through TCP.
+              cmds=('dig -t A @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
+                    'dig -t AAAA @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
+                    'dig -t A +tcp @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
+                    'dig -t AAAA +tcp @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"')
+              declare -p cmds
+              for i in ${!cmds[*]}
+              do
+                ips=($(eval "${cmds[i]}"))
```

Comment:
Why not just
```bash
ips=($("${cmd}"))
```
?

Reply:
It won't resolve vars:

```
[mdulko@dulek-workstation kuryr-kubernetes]$ for cmd in "${cmds[@]}"; do $("${cmd}"); done
bash: dig -t A @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}": command not found...
bash: dig -t AAAA @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}": command not found...
bash: dig -t A +tcp @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}": command not found...
bash: dig -t AAAA +tcp @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}": command not found...
[mdulko@dulek-workstation kuryr-kubernetes]$ for cmd in "${cmds[@]}"; do $(${cmd}); done
dig: couldn't get address for '"${NAMESERVER}"': not found
dig: couldn't get address for '"${NAMESERVER}"': not found
dig: couldn't get address for '"${NAMESERVER}"': not found
dig: couldn't get address for '"${NAMESERVER}"': not found
```

Reply:
That's because you are using single quotes in `cmds` array :) The can be unfloded during array declaration. I.e.
```bash
cmds = ('dig -t A @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
...
```
can be replaced by
```bash
cmds = ("dig -t A @${NAMESERVER} +short ${svc}.${CLUSTER_DOMAIN}"
...
```

Reply:
Seems like it works, but only if I use `ips=($(${cmd}))` (no quotes). I can change it though I'm not sure why it was escaped like this in the first place.

Reply:
I prefer to leave the quoting around the parameter expansions inside the commands.

Reply:
If you want to get rid of the subscripting, the following should work:
```bash
    cmds=('dig -t A @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
          'dig -t AAAA @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
          'dig -t A +tcp @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"'
          'dig -t AAAA +tcp @"${NAMESERVER}" +short "${svc}.${CLUSTER_DOMAIN}"')
    for cmd in "${cmds[@]}"
    do
        ips=($(eval "$cmd"))
        # ...
```